### ISSUE TITLE
E2EE Requirements

### ISSUE THREAD

**Issue by insertish** [revoltchat/revolt#207]
MVP List:
- OMEMO-style secret chats
- key verification (verify users through trusted channel, can use emojis like Matrix)
- server key storage (generate a mnemonic to encrypt and store keys on server)

Additional Requirements:
- Cross device signing (i.e. https://element.io/blog/e2e-encryption-by-default-cross-signing-is-here/)
- New device trust flow (allow a trusted device to confirm a new session)
- Expiring / ephemeral messages
- E2EE Voice Calls
- E2EE File Uploads

**Comment by Flam3z** [comment]
I think i'd be best to have e2ee by default since alot of people are also unaware of the perks it can have on your privacy and security. Or at least give the user a small tour on what e2ee is and have them decide if they want it for all DMs. I do agree with the telegram DM deletion method as it helps delete dms easier without worry of the other user snooping through old history.

**Comment by UnixPhonez** [comment]
E2EE is a must for DMs! at least

**Comment by Flam3z** [comment]
> E2EE is a must for DMs! at least and by default

Agreed, E2EE for servers is redundant as servers are public. DMs, groups and notes should be E2EE.

**Comment by peepopoggers** [comment]
E2EE should be enabled by default in dms (and groups hopefully later), and if there is any reason to not have it enabled by default, then it has been implemented wrong.

**Comment by erkinalp** [comment]
@insertish MLS is stronger than olm that is used by OMEMO and MegOLM.

**Comment by Flam3z** [comment]
Question, if they keys are staying on the server wouldn't that garner fear that revolt devs can just take any keys and look into coversations at any time? If I am wrong clarification is appreciated.

**Comment by Zomatree** [comment]
@Flam3z The private key wouldnt be stored on the server, so no.

**Comment by Flam3z** [comment]
Ty for the quick response

**Comment by Evilhotdog** [comment]
> Question, if they keys are staying on the server wouldn't that garner fear that revolt devs can just take any keys and look into coversations at any time? If I am wrong clarification is appreciated.

No, the server would receive the public keys of each user but only the users would have the full information needed to decrypt

**Comment by ggtylerr** [comment]
> > E2EE is a must for DMs! at least and by default
> 
> Agreed, E2EE for servers is redundant as servers are public. DMs, groups and notes should be E2EE.

...Not all servers are public, what about private ones?

**Comment by Flam3z** [comment]
> > > E2EE is a must for DMs! at least and by default
> > 
> > 
> > Agreed, E2EE for servers is redundant as servers are public. DMs, groups and notes should be E2EE.
> 
> ...Not all servers are public, what about private ones?

that could possibly be a good feature but that depends if group chats are only limited to 10 users. If its considered, i think a message should appear to the owner explaining its possible draw backs and pros (i.e. server unable to be posted on the discovery but the main revolt server are unable to see your meesages)

**Comment by peepopoggers** [comment]
Implementing encryption for a server with a large amount of people is a bigger ordeal than doing the same for dms or groups. For dms, it is merely about exchanging keys and then just using them when sending messages. With groups, it is a little different, but for large servers I think a different protocol would be needed to add encryption. If servers had like 20 members for example, it would be easy, but if they had 1000+, then you would have to be careful not to stress any hardware too much with the protocol used. Adding encryption to dms and groups would definitely be a very strong point to revolt however and the platform would gain more support from the privacy community.

**Comment by Flam3z** [comment]
agreed, as the point of servers are "public" by design and if you want a private group then thats what the group dms are for (depending on the group user limit if there is one) e2ee groups don't really make sense in practice.

**Comment by ggtylerr** [comment]
> > ...
> 
> that could possibly be a good feature but that depends if group chats are only limited to 10 users. If its considered, i think a message should appear to the owner explaining its possible draw backs and pros (i.e. server unable to be posted on the discovery but the main revolt server are unable to see your meesages)

>agreed, as the point of servers are "public" by design and if you want a private group then thats what the group dms are for (depending on the group user limit if there is one) e2ee groups don't really make sense in practice.

Even if this is the case, there's multiple reasons why a small group would want a server instead - Voice calls don't alert everyone in the group, multiple channels, custom emojis (if and when that ever becomes a thing), permissions, etc.

I do agree with @peepopoggers though and this should be limited to 30, maybe 50 members max.

**Comment by peepopoggers** [comment]
Yeah, you are right. Idk the technical limitations of this stuff. For example, would e2ee be possible for servers with 5000, or even 50,000 members?

**Comment by Flam3z** [comment]
i'd say e2ee shall be in dms, groups and notes. Groups should at least be able to hold 20 people. But at the end of the day it all depends on the hardware and money they have.

**Comment by erkinalp** [comment]
@peepopoggers Yes, Matrix for example has E2EE rooms with more than 50k members. And underlying key sharing protocols support arbitrarly large groups. However, the larger the group, the more keyshare messages you will need.

**Comment by Flam3z** [comment]
> @peepopoggers Yes, Matrix for example has E2EE rooms with more than 50k members. And underlying key sharing protocols support arbitrarly large groups. However, the larger the group, the more keyshare messages you will need.

yes but encrypted rooms take about a minute to send messages and idk if its their servers or the protocol itself.

**Comment by peepopoggers** [comment]
> i'd say e2ee shall be in dms, groups and notes. Groups should at least be able to hold 20 people. But at the end of the day it all depends on the hardware and money they have.

Make a priority for groups, dms and notes, and servers could be experimental until hopefully refined to a well working state.

**Comment by peepopoggers** [comment]
> > @peepopoggers Yes, Matrix for example has E2EE rooms with more than 50k members. And underlying key sharing protocols support arbitrarly large groups. However, the larger the group, the more keyshare messages you will need.
> 
> yes but encrypted rooms take about a minute to send messages and idk if its their servers or the protocol itself.

I think it would be cool to implement it like the way I said above, experimental and then it can be optimized hopefully.

**Comment by Flam3z** [comment]
we'll see when the milestone gets there

**Comment by UnixPhonez** [comment]
this will bring more users

**Comment by UnixPhonez** [comment]
I would like to add encryption via DM's should be mandatory meaning it's automatically added without opt out

**Comment by FillingTheVoidOnYT** [comment]
Encryption should be prioritized...

**Comment by insertish** [comment]
> Encryption should be prioritized...

I think there are more important things we should address currently, like getting core features in that people expect from other platforms.

I'm also not going to rush building encryption either because there's only one chance to get it right, and I don't want to get it wrong.

**Comment by insertish** [comment]
> I would like to add encryption via DM's should be mandatory meaning it's automatically added without opt out

I don't know how the final implementation will look, but I'm personally leaning towards having a toggle between the two which also allows us to:
- Keep old DM contents
- Allow users to have message history on server (one possibility with encrypted messages is to do an encrypted server-side backup but it can introduce holes or potentially lose data, not to mention that all messages would need to be merged and all of that data needs to be saved locally)

**Comment by FillingTheVoidOnYT** [comment]
> > I would like to add encryption via DM's should be mandatory meaning it's automatically added without opt out
> 
> I don't know how the final implementation will look, but I'm personally leaning towards having a toggle between the two which also allows us to:
> 
> * Keep old DM contents
> * Allow users to have message history on server (one possibility with encrypted messages is to do an encrypted server-side backup but it can introduce holes or potentially lose data, not to mention that all messages would need to be merged and all of that data needs to be saved locally)

okay understandable, but let me ask you this when messages are purged in an channel are they deleted on the server side or just kept in plain text despite being deleted in their channel they correspond to

**Comment by erkinalp** [comment]
@insertish Could a Matrix-style encapsulation would work? Matrix encrypts E2EE message event body and wraps it in a whole event body. That would allow enabling encryption from a certain point on, without losing the ability to read messages prior to enabling E2EE.

**Comment by insertish** [comment]
> > > I would like to add encryption via DM's should be mandatory meaning it's automatically added without opt out
> > 
> > I don't know how the final implementation will look, but I'm personally leaning towards having a toggle between the two which also allows us to:
> > 
> > * Keep old DM contents
> > * Allow users to have message history on server (one possibility with encrypted messages is to do an encrypted server-side backup but it can introduce holes or potentially lose data, not to mention that all messages would need to be merged and all of that data needs to be saved locally)
> 
> okay understandable, but let me ask you this when messages are purged in an channel are they deleted on the server side or just kept in plain text despite being deleted in their channel they correspond to 

Messages have always and are deleted completely and immediately, with the exception of attachments which can take up to a minute to delete because the file server needs to delete it from S3.

**Comment by therealonlygithub1** [comment]
> > E2EE is a must for DMs! at least and by default
> 
> Agreed, E2EE for servers is redundant as servers are public. DMs, groups and notes should be E2EE.

although it's redundant for public servers, there should still be the option for servers in general as not all servers are public. Many servers would like the encryption part it would actually bring more companies to revolt as many companies operate under an NDA contract and encryption would be one less checkmark they don't have to leave unchecked

**Comment by erkinalp** [comment]
> that the services is covered against breach of law they don't even aware of.

Can you explain it further with references to actual legal decisions supporting this?

**Comment by go-compile** [comment]
E2EE would be great, however it will require a **upgrade to user authentication**.

Currently, Revolt sends the password to the server, if the current authentication system was to be used with E2EE this would compromise all value added by E2EE. This is because the public/private key pairs will be encrypted on the server using a key derived from the user's password. Hence fourth, without a authentication overall E2EE would be **vulnerable to MiTM attacks and a malicious server** exposing keys to retrieve plaintext communications.

## Solution
Authentication will need to use a Zero Knowledge Proof (ZKP), this would project the password from both MiTM attacks and a malicious servers. A augmented PAKE (Password Authenticated Key Exchange) - would allow for the server to prove the user knows his password **without** ever seeing his password, not even in registration.

A scheme such as **OPAQUE** (<https://blog.cloudflare.com/opaque-oblivious-passwords/>) or Secure Remote Password protocol  (**SRP**) would be the best choice for this. Moreover, these schemes are used by Protonmail (https://proton.me/blog/encrypted-email-authentication), Tutanota etc. Additionally, further key derivation via the use of a KDF, e.g. HKDF, should be used to produce two unique keys, one for authentication, the other for encrypting the private key.

**Comment by MasicoreLord** [comment]
Scared how proposed laws like the current Earn It Act (USA), the Online Safety Bill (UK) or etc. well impact this, assuming they pass eventually.

**Comment by insertish** [comment]
Just making a note here that this will need to be converted to an RFC at some point before further work can be done.

> E2EE would be great, however it will require a **upgrade to user authentication**.

Definitely, I will add this to our internal queue to look into.

> Scared how proposed laws like the current Earn It Act (USA), the Online Safety Bill (UK) or etc. well impact this, assuming they pass eventually.

If passed, that's just more motivation to work on this, I will never recognise any law that bans / hampers encryption to a degree that it is useless.

**Comment by MasicoreLord** [comment]
> Just making a note here that this will need to be converted to an RFC at some point before further work can be done.
> 
> > E2EE would be great, however it will require a **upgrade to user authentication**.
> 
> Definitely, I will add this to our internal queue to look into.
> 
> > Scared how proposed laws like the current Earn It Act (USA), the Online Safety Bill (UK) or etc. well impact this, assuming they pass eventually.
> 
> If passed, that's just more motivation to work on this, I will never recognise any law that bans / hampers encryption to a degree that it is useless.

While true, in countries like the USA for example, you'd risk losing liability protections and be treated as a publisher rather than a platform, liable for anything you happen to have on your service.

**Comment by MasicoreLord** [comment]
Since the main instance isn't USA based it may be safe from that particular threat.

**Comment by Pantyhose-X** [comment]
Single channel and DM encryption should be mandatory, and encrypted channels should have colored names by default.
Creating a single channel encryption is better than encrypting the entire server
Worthless image and video files should not be sent to encrypted channels

**Comment by vertigo220** [comment]
> Worthless image and video files should not be sent to encrypted channels

Not sure if you meant all image and video files are worthless and not necessary to encrypt or that specifically ones that are worthless don't need it, but if the former, that's simply not true, and if the latter, it becomes an issue of how to tell between ones that do and do not need it and running the risk of accidentally not encrypting ones that should be. Encryption isn't exactly an expensive operation, so I see no reason to not just encrypt everything.

**Comment by c0fe** [comment]
Is there interest in implementing Signal protocol as the encryption? It can be adjusted to be used without phone numbers, it is proven to be secure unlike Matrix, supports media, video chatting and group chat.

**Comment by Pantyhose-X** [comment]
Clients may implement E2EE in the future, but servers will never implement E2EE, I've already asked them at revolt.
> Using all 3 services Revolt, Lightspeed, & Gifbox is giving Revolt a lot of your data. Your chats, images, & videos all logged stored for whatever purpose they see fit. Lets go back to the site & talk about the part of the slogan where it says "you can be sure your conversations are confidential & data is secure." Revolt doesn't have E2EE (end to end encryption) so confidential is a lie because everyone's chat is stored & any staff member can read from the beginning to the present.

> They advertised Revolt had E2EE in it's first year which was a lie it was on a table of discussions to be implemented in future rewrite of the client. They spend more time playing video games, making jokes condescending America, sidetracking on multiple projects simultaneously, generating memes with Stable Diffusion, & doing school work then working on the software.

> When it comes to self hosting Revolt they did everything to deter people from self hosting because they didn't want other people running their own instance. They wouldn't support anyone who asked for help which goes against the open source license policy as you must provide instructions & support for your open software your distributing. If anyone asked you were interrogated for asking they were afraid people would have no reason to go to Revolt.chat if they help people setup theirs.
There were also certain parts of Revolt that were closed-sourced like server discovery & more they did this to ensure people would not have a full functioning chat service ready to go & you would have to do some manual implementations such as creating your own server discovery & making it work within the app. They did this on purpose to spitefully stop people from self-hosting who wants a discord clone with no server discovery & you have to make one yourself?

> The staff is LGBTQIA so anything in that direction = instant ban but condescending America is alright 100%. You could even make your account username as Ihateamericanscum & guess what you wouldn't get banned or forced to change your username but if your name was Ihateeurope or LGBT you would be immediately banned or forced to change username before you got to say hello. They also have a serious hatred for the AM|PM format I don't understand why there is such a great hatred for it that they have purposely refused to implement it & argue with people to just learn 24/7 time format.

> Some Criticism Source https://0.0g.gg/?f47f5f63dd027f2f#85cGkzR2V7z4R1t6agghesCvxsY9DicwDpQ6m14Wq5KD

**Comment by Zomatree** [comment]
> Clients may implement E2EE in the future, but servers will never implement E2EE, I've already asked them at revolt.

This is complete miss-information, we have always planned and still plan on implementing end to end encryption fully into revolt.

**Comment by insertish** [comment]
There's a lot more nuance here that people miss, it's not that servers can't be encrypted, it's that it may be technically infeasible (very very large servers reaching thousands of users) or useless in some cases (servers with public invites).
The simple answer is no, but the actual answer is that work needs to be done to figure out how it'll actually work within servers because it's not as simple as DMs.

**Comment by BirdInFire** [comment]
> There's a lot more nuance here that people miss, it's not that servers can't be encrypted, it's that it may be technically infeasible (very very large servers reaching thousands of users) or useless in some cases (servers with public invites). The simple answer is no, but the actual answer is that work needs to be done to figure out how it'll actually work within servers because it's not as simple as DMs.

but why not starting by implementing a "simple" DM, and follow with a double method, one based on small and medium server and later push very large, because, 90% of your user's server will fall into small and medium.

**Comment by Rexogamer** [comment]
I don't really see how that addresses the issues Insert raised..?

**Comment by BirdInFire** [comment]
> I don't really see how that addresses the issues Insert raised..?

it's help to do the requirement step by step, by starting by the "easiest" (1 to 1 discussion), and continue after, because here many ask THE solution that work everywhere, and we see in Whatsapp / Signal case, that this "one solution" do not exist yet, so if you don't start simple you will never finish this work.

**Comment by erkinalp** [comment]
Whatsapp and Signal do client fanout with pairwise channels, not true group key establishment.

**Comment by BirdInFire** [comment]
> Whatsapp and Signal do client fanout with pairwise channels, not true group key establishment.

Was an example to make understand my point not really was accurate, I can use element if you prefer

**Comment by Neustradamus** [comment]
Any progress on this ticket?

**Comment by c0fe** [comment]
> > Whatsapp and Signal do client fanout with pairwise channels, not true group key establishment.
> 
> Was an example to make understand my point not really was accurate, I can use element if you prefer

Matrix isn't very good security wise, Signal is significantly better. Matrix is right up next to plaintext, while it isn't nothing it is pretty close to nothing in terms of security/privacy.

**Comment by c0fe** [comment]
@insertish I would suggest using Signal protocol for this.

**Comment by erkinalp** [comment]
@c0fe That protocol does not scale well.

**Comment by c0fe** [comment]
@erkinalp can you please expand on this? where have you seen it take issue with scaling?

**Comment by c0fe** [comment]
@erkinalp the limits as far as I am aware of is at a 1,000 at the moment. Last I heard was that there was work towards increasing that. Yeah that's a limit, but it still far better to offer that than something that is inherently broken in terms of security.

**Comment by vertigo220** [comment]
@c0fe You keep referring to Matrix as broken, etc, but I see no evidence of that. I looked into it pretty extensively a while back, and again a bit the other day based on your comment, and while it's not audited to the same extent as Signal and does have some issues, AFAICT it's *far* from "broken" or "next to plaintext." Care to elaborate?

**Comment by BirdInFire** [comment]
> @c0fe You keep referring to Matrix as broken, etc, but I see no evidence of that. I looked into it pretty extensively a while back, and again a bit the other day based on your comment, and while it's not audited to the same extent as Signal and does have some issues, AFAICT it's *far* from "broken" or "next to plaintext." Care to elaborate?

Actually matrix is quite broken in fact that it happen so often that group encryption broke and need manual intervention to fix so it's not quite suitable for the use in revolt

**Comment by c0fe** [comment]
@vertigo220 the problems with Matrix protocol is that there are numerous security problems that are found by researchers. We aren't even talking about one big issue, say Specter we are talking multiple of these issues being found in terms of how the implementation is done, encryption/decryption and so forth.

In other words Matrix protocol is rich in vulnerabilities, people find lots of them and they are all high severity.

**Comment by vertigo220** [comment]
Again, I've looked into this, and these statements seem exaggerated. There were a handful of flaws found in 2022 which were responsibly disclosed and, AFAICT, handled appropriately by the developers and fixed in a reasonable time. There was one in 2021 that was a flaw with some clients and *not* the protocol and that was fixed. And there was an attack in 2019 where malicious actors gained access to _unencrypted_ messages and to password hashes and access tokens which, while far from ideal, isn't really that big a deal if you use a decent password, and I don't see anything about any attacks since. So yeah, while it's had some issues, nothing jumps out as being all that terrible and I'd say it's far from broken or "rich in vulnerabilities" with people findings "lots of them." And of course it requires manual intervention to fix them, but the key is they're being fixed.

Sure, as I said before, Signal has been more thoroughly audited and has a better track record, but it's not perfect, either, and IIRC vulnerabilities have been found with it as well (I highly doubt it hasn't had any). If perfect were a 10, Signal might be an 8 and Matrix a 6, whereas Discord is a 0. So between using the Signal protocol and Matrix protocol, as far as security is concerned, sure, go with Signal, but if it has limitations and doesn't provide useful features the Matrix protocol does, I don' think it's necessary to avoid using Matrix, as it's still very good security-wise, and Revolt, being intended as a Discord replacement (i.e. a casual chat client mainly for gaming) needs to make sure it can do that job more than it needs to ensure the most bulletproof encryption possible. People looking for that can just use Signal.

And I certainly don't think making statements that Matrix is totally broken and riddled with issues is accurate or fair, to Matrix itself or to users looking for something to use. While not perfect, and not something that should probably be used to transmit highly sensitive info or by someone looking to hide from governments, I'd say it's plenty good enough to maintain privacy. Unless I'm missing something, but I did ask for elaboration and didn't receive any specifics, and I haven't been able to find anything on my own so far.

**Comment by BirdInFire** [comment]
> Again, I've looked into this, and these statements seem exaggerated. There were a handful of flaws found in 2022 which were responsibly disclosed and, AFAICT, handled appropriately by the developers and fixed in a reasonable time. There was one in 2021 that was a flaw with some clients and *not* the protocol and that was fixed. And there was an attack in 2019 where malicious actors gained access to _unencrypted_ messages and to password hashes and access tokens which, while far from ideal, isn't really that big a deal if you use a decent password, and I don't see anything about any attacks since. So yeah, while it's had some issues, nothing jumps out as being all that terrible and I'd say it's far from broken or "rich in vulnerabilities" with people findings "lots of them." And of course it requires manual intervention to fix them, but the key is they're being fixed.
> 
> 
> 
> Sure, as I said before, Signal has been more thoroughly audited and has a better track record, but it's not perfect, either, and IIRC vulnerabilities have been found with it as well (I highly doubt it hasn't had any). If perfect were a 10, Signal might be an 8 and Matrix a 6, whereas Discord is a 0. So between using the Signal protocol and Matrix protocol, as far as security is concerned, sure, go with Signal, but if it has limitations and doesn't provide useful features the Matrix protocol does, I don' think it's necessary to avoid using Matrix, as it's still very good security-wise, and Revolt, being intended as a Discord replacement (i.e. a casual chat client mainly for gaming) needs to make sure it can do that job more than it needs to ensure the most bulletproof encryption possible. People looking for that can just use Signal.
> 
> 
> 
> And I certainly don't think making statements that Matrix is totally broken and riddled with issues is accurate or fair, to Matrix itself or to users looking for something to use. While not perfect, and not something that should probably be used to transmit highly sensitive info or by someone looking to hide from governments, I'd say it's plenty good enough to maintain privacy. Unless I'm missing something, but I did ask for elaboration and didn't receive any specifics, and I haven't been able to find anything on my own so far.

You actually never used matrix it seem because having to restart a new group because the encrusted have broken everything is an horror, and debug one user who is the only one to have problems is annoying as hell for me me matrix protocol is far from being stable enough to be used by revolt

**Comment by vertigo220** [comment]
> You actually never used matrix it seem because having to restart a new group because the encryption have broken everything is an horror, and debug one user who is the only one to have problems is annoying as hell for me me matrix protocol is far from being stable enough to be used by revolt

Okay, so now your true motives/feelings have come to light. Actually, yes, I've been using it for the past year or so (and after trying, and being disappointed by, Revolt, BTW), and certainly not without issues, but overall it's worked quite well, actually better than Signal did when I used it. I'm not particularly for or against either, I just know there aren't a lot of options out there, and Matrix and Signal both have their pros and cons, and personally I prefer Matrix for usability and, while probably not quite as secure as Signal, it's secure enough. You, on the other hand, seem to dislike it because of issues you've experienced while using it, and you're turning that into arguing that it's not secure, which are two different things.

Matrix has a long way to go, but frankly so does Signal, and I don't think Matrix is completely broken, certainly not with respect to its encryption, and I've seen nothing here to support that idea, just generalized talk with no specifics, followed by assumptive statements like this. I just want a secure, usable protocol that I can use to chat with and call friends and family, and after trying and giving up on Signal, Discord, Revolt, Guilded, and a few others, I've found Matrix to be the best fit for that. As I've said, it's not perfect by any means, and I don't have any personal attachment/connection to it and would be very interested to know if there are any significant issues with it (hence my asking for clarification on such suggestive comments here), but absent any actual evidence, and based on my own research, Matrix seems plenty adequate for my and most people's needs, and certainly far more private and secure than Revolt. I'm not against implementing the Signal protocol into Revolt (or any other platform), as it is absolutely a secure protocol, but a) that's only if it can perform the functions required, which I question if it could in this case, and b) that security can only be assured if the platform, and the Signal protocol's implementation, were fully open, which doesn't appear to be the case with Revolt.

I'm honestly confused by people that on the one hand criticize (without offering _any_ specifics or sources) the Matrix protocol, yet on the other use a program that isn't even fully open-source and has even less auditing than Matrix. You guys criticize Matrix because of flaws found--despite the fact that _any_ software, be it Signal, linux, Android, or even DARPA's, has had vulnerabilities that have been discovered with enough testing--yet at least Matrix is prominent enough to have security researchers and others actually look for flaws (which are found and fixed), whereas the only reason Revolt hasn't had flaws found is likely because it doesn't receive the same attention. And before you make another (wrong) assumption, I'm not anti-Revolt, pro-Matrix, or anything else (except anti-Discord); I just want a truly secure and private messaging platform that works well. IME, Signal is secure but not usable and not all that private, Revolt is not secure, private, or usable, and Matrix is at least moderately secure and private and mostly usable, certainly more so than the others.

**Comment by DeclanChidlow** [comment]
Just figured I'd jump in and say that E2EE is still in a very early stage and minimal discussion has taken place regarding it. In the event that it's added there would be an RFC first to discuss implementation.

Revolt also doesn't plan to use any existing protocols as it would require changes to the way the platform is used on a much more general level. Revolt doesn't intend to be a Signal or Element client and from all I've seen of the protocols they aren't hugely flexible in the way Revolt would need.

Also, I can see this discussion potentially taking a hostile turn, so remember to act respectfully.

**Comment by vertigo220** [comment]
> Just figured I'd jump in and say that E2EE is still in a very early stage and minimal discussion has taken place regarding it. In the event that it's added there would be an RFC first to discuss implementation.
> 
> Revolt also doesn't plan to use any existing protocols as it would require changes to the way the platform is used on a much more general level. Revolt doesn't intend to be a Signal or Element client and from all I've seen of the protocols they aren't hugely flexible in the way Revolt would need.
> 
> Also, I can see this discussion potentially taking a hostile turn, so remember to act respectfully.

To clarify, Element isn't the protocol, it's a UI/application that uses the Matrix protocol.

As for Revolt using its own vs an existing one, that's highly recommended against, and it would make Revolt considerably less secure, at least in theory. So any comparison between Signal, Matrix, and whatever Revolt were to use would place Revolt far below the other two, so anyone, here and elsewhere, that considers Matrix unsecure would have to consider Revolt even less secure. And yes, I'm aware Matrix is in the same boat, having rolled their own, but it's been around longer, is much more widespread/popular, and therefore has been scrutinized much more.

I want to be clear: I'm not saying Matrix is as secure as Signal or that Signal shouldn't be used. I'm simply saying that, from everything I've read, Matrix is far from unsecure, and certainly not to the degree to refer to it as "broken" or "next to plaintext" or "rich in vulnerabilities." It shouldn't be discounted based on security, especially if it can fulfill the role more than the Signal protocol can. If that's _not_ the case, I'd _love_ to know, because I use it. But I see no evidence of that, from my own research or from anything offered by those criticizing it, apparently baselessly, here. I just take issue with people turning others away from something that's perfectly adequate for most users' needs, especially when they then turn around and support a project that's even worse. That just really comes off as a personal grievance against the former or false support for the latter, and that's not helping people looking for something to use.

So back to this comment, if Revolt won't use either, regardless of which is "better," that makes the argument and choices simpler, as it's then a choice between a longstanding, proven, heavily audited, but limited usability protocol (Signal), a longstanding, mostly proven, moderately audited, more usable protocol (Matrix), and a new, unproven, unaudited protocol of unknown usability created by people that, AFAICT, have zero encryption knowledge (I could be wrong, and I don't mean that as a slam, just stating things as I see them, and the fact someone involved in Revolt doesn't seem to know the difference between Element and Matrix reinforces that). That, to me, makes Revolt potentially fine as a pure Discord replacement, where privacy doesn't matter, but for standard chat/calls/etc, where it does, it makes it a no-go, which leaves Matrix or Signal. And frankly, IMO, it would be better in multiple respects to just not bother with E2EE in that case, and spend the resources on making it a better client for the intended purpose, instead of wasting those resources creating a potentially subpar implementation that is not in the main scope of the project and only serves to offer a false sense of security to users. But really, I don't even care if Revolt uses Signal, Matrix, rolls its own, or doesn't do it at all, because Matrix works fine for me, is secure (enough), and is fully open-source. My main concern was people proclaiming Matrix as broken, and I just wanted to know, since I use it, if that was true, and I've seen no evidence offered that it is, which then concerns me because that suggests people are making false statements and misleading others. If I'm wrong, fine. All I ask is for proof, because I truly want to know if those statements are factual.

**Comment by DeclanChidlow** [comment]
I'm aware Element is a client. Merely a slip up, my apologies.

**Comment by vertigo220** [comment]
> I'm aware Element is a client. Merely a slip up, my apologies.

No worries, easily mixed up, just wanted to keep things clear.

**Comment by BirdInFire** [comment]
> > You actually never used matrix it seem because having to restart a new group because the encryption have broken everything is an horror, and debug one user who is the only one to have problems is annoying as hell for me me matrix protocol is far from being stable enough to be used by revolt
> 
> 
> 
> Okay, so now your true motives/feelings have come to light. Actually, yes, I've been using it for the past year or so (and after trying, and being disappointed by, Revolt, BTW), and certainly not without issues, but overall it's worked quite well, actually better than Signal did when I used it. I'm not particularly for or against either, I just know there aren't a lot of options out there, and Matrix and Signal both have their pros and cons, and personally I prefer Matrix for usability and, while probably not quite as secure as Signal, it's secure enough. You, on the other hand, seem to dislike it because of issues you've experienced while using it, and you're turning that into arguing that it's not secure, which are two different things.
> 
> 
> 
> Matrix has a long way to go, but frankly so does Signal, and I don't think Matrix is completely broken, certainly not with respect to its encryption, and I've seen nothing here to support that idea, just generalized talk with no specifics, followed by assumptive statements like this. I just want a secure, usable protocol that I can use to chat with and call friends and family, and after trying and giving up on Signal, Discord, Revolt, Guilded, and a few others, I've found Matrix to be the best fit for that. As I've said, it's not perfect by any means, and I don't have any personal attachment/connection to it and would be very interested to know if there are any significant issues with it (hence my asking for clarification on such suggestive comments here), but absent any actual evidence, and based on my own research, Matrix seems plenty adequate for my and most people's needs, and certainly far more private and secure than Revolt. I'm not against implementing the Signal protocol into Revolt (or any other platform), as it is absolutely a secure protocol, but a) that's only if it can perform the functions required, which I question if it could in this case, and b) that security can only be assured if the platform, and the Signal protocol's implementation, were fully open, which doesn't appear to be the case with Revolt.
> 
> 
> 
> I'm honestly confused by people that on the one hand criticize (without offering _any_ specifics or sources) the Matrix protocol, yet on the other use a program that isn't even fully open-source and has even less auditing than Matrix. You guys criticize Matrix because of flaws found--despite the fact that _any_ software, be it Signal, linux, Android, or even DARPA's, has had vulnerabilities that have been discovered with enough testing--yet at least Matrix is prominent enough to have security researchers and others actually look for flaws (which are found and fixed), whereas the only reason Revolt hasn't had flaws found is likely because it doesn't receive the same attention. And before you make another (wrong) assumption, I'm not anti-Revolt, pro-Matrix, or anything else (except anti-Discord); I just want a truly secure and private messaging platform that works well. IME, Signal is secure but not usable and not all that private, Revolt is not secure, private, or usable, and Matrix is at least moderately secure and private and mostly usable, certainly more so than the others.

Think what you want I maintain the fact that any other encryption method is better for revolt than matrix one

**Comment by erkinalp** [comment]
@insertish Important nit while implementing MLS: force the group key tree of each group to have leaf nodes per-user. And then have each user have per-device key distribution tree, with the master key of the per-user tree equalling the leaf key of the user from the group tree. That way, you would not leak one's devices.

**Comment by c0fe** [comment]
@DeclanChidlow are you suggesting Revolt roll their own encryption protocol? I would strongly advise against that as we can have another Matrix-esque type of situation. Encryption is very difficult to get right and so rolling your own is just creating more problems for yourself.

I am not suggesting in becoming another Signal client, you can use the Signal protocol without just being a Signal client. For example WhatsApp implements it as well as Google for their Google Messages which was previously Allo and Microsoft for Skype.

**Comment by noahbliss** [comment]
Potentially unhelpful/low context comment, but just throwing out there that the Keybase approach could be a viable idea too. Keybase had some really cool fundamental design angles, just unfortunately more or less dead thanks to being bought by zoom. It's open source though.

**Comment by isotopp** [comment]
Unpopular Opinion as a comment: (from https://infosec.exchange/@isotopp/114697563169505268)

# Not being federated and E2E as an advantage

https://revolt.chat/ is a chat that ultimately tries to emulate and displace Discord as an alternative.

Revolt is not E2E encrypted and not federated. That is a big plus.

Federation and E2E encryption are features that make an implementation much, much more complicated without need. Most people do not want to organize a women's shelter or abortion support or the antifash resistance. They just want to have a public guild server for their gaming guild or chat about 3D printers.

# Don't federate chat

Having a federated system for that has a number of implications that make implementations more complicated and impact quality of life for users.

In a federated systems, different servers can be on different versions of the server software. This can impact features, which may or may not work properly. If will affect security of the whole network due to lack of patches in one place. And it will create propagation delays of messages inside the federated network, that is, order or even availability of messages can differ depending on what server in the federated network you are on.

As a consequence for the individual user, there is no single history order of messages – some messages may be missing, appear out of order, and there will be spam. Lots of it.

Back when Google Hangouts was still federating XMPP (Jabber) servers, specialized Jabber server implementations existed with the sole purpose of simulating user profiles, discovering Hangouts that are joinable, and spamming users with, well, garbage.

Operating a public chat room in 2025 is already a moderation nightmare. Adding federation is a surefire way to make things fail.

## Federate identity, multihome clients.

This is not to say that the client should be bound to a single server. 

It can, and should, in fact, connect to different servers for each community. The server,  though, should be a "single visible instance, single history for all" thing, and have proper controls for the moderation team to control admission of new users to whatever their policy is.

**EDIT:** And, obviously, federated identity can be quite useful. 

In fact, even Mastodon and other Fediverse systems should use federated identity, and also be identity servers. This will allow me to connect my client to a chat server without making any new account.

# Don't E2E encrypt chat

Use TLS (or even a REST protocol over HTTPS) to connect to the chat server. But do not E2E messages.

Keep a clear text history on the server.

This makes it trivial for a single identity ("isotopp") to have multiple devices (desktop, ipad and web client) and switch between them, yet have the same chat history on all devices. 

It will allow new users to join a channel and see the full channel history, if the channel is configured to do that (and because the server is not federated, the channel backlog policy can actually be enforced).

It makes it trivial to implement search and indexed archives for a server that wants that.

It makes it easier to implement detection of spammy behavior, implement rate limits, and to autodetect inappropriate or banned content.

With E2E, the server only has binary garbage that it cannot read itself, only distribute. It may not even have metadata.

It will be complicated to scan messages for spammy behavior, detect banned content.

It will be impossible to provide backlog to newly joined users. An existing identity (Isotopp) adding a new device (his phone) can only have backlog, if the other clients (the desktop of ipad client of Isotopp) of that user support client-to-client backlog transfer.

Without E2E encryption, there is no keysplit problem ("half a channel only sees undecipherable messages"), there is no "known plaintext" problem to handle, and there is no large-scale key distribution problem.

# Conclusion

That is, a project that

- uses TLS, but not E2E, and
- that uses multi-homed clients but not federation,

will be able to build a better client with better UX with less effort and be a lot more agile. 

Both properties (federation, E2E) do little for the end user experience except in a few limited use-cases, but come with a huge cost in implementation complexity, review and anti-abuse measures that will ultimately bind the majority of your dev capacity for nothing in return.

**Comment by imsys** [comment]
Aside of the Matrix protocol, there is also the SimpleX protocol ( https://github.com/simplex-chat/SimpleX-chat ). Maybe there are things that could be borrowed from them.

**Comment by Zomatree** [comment]
We don't plan on stopping implementing end to end encryption.
